python-checks/src/main/java/org/sonar/python/checks/PandasChainInstructionCheck.java

```diff
@@ -0,0 +1,136 @@
+import org.sonar.plugins.python.api.symbols.FunctionSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.SubscriptionExpression;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+import org.sonar.python.types.InferredTypes;
+
+@Rule(key = "S6742")
+public class PandasChainInstructionCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Refactor this long chain of instructions with pandas.pipe";
+  private static final int MAX_CHAIN_LENGTH = 5;
+
+  private static final String DATAFRAME_FQN = "pandas.core.frame.DataFrame";
+
+  private final Set<QualifiedExpression> visited = new HashSet<>();
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> visited.clear());
+    context.registerSyntaxNodeConsumer(Tree.Kind.QUALIFIED_EXPR, this::checkChainedInstructions);
+  }
+
+
+  private void checkChainedInstructions(SubscriptionContext ctx) {
+    QualifiedExpression expr = (QualifiedExpression) ctx.syntaxNode();
+    List<QualifiedExpression> chain = visitParent(expr, new ArrayList<>());
+    if (chain.size() >= MAX_CHAIN_LENGTH && isValidPandasCall(chain)) {
+      ctx.addIssue(chain.iterator().next(), MESSAGE);
+    }
+  }
+
+  private List<QualifiedExpression> visitParent(QualifiedExpression expr, List<QualifiedExpression> chain) {
+    if (visited.contains(expr)) {
+      return chain;
+    }
+    visited.add(expr);
+    chain.add(expr);
+
+    if (expr.qualifier().is(Tree.Kind.CALL_EXPR)) {
+      return visitCalleeParent((CallExpression) expr.qualifier(), chain);
+    } else if (expr.qualifier().is(Tree.Kind.SUBSCRIPTION)) {
+      return ignoreSubscriptionAndGetCallExpr((SubscriptionExpression) expr.qualifier())
+        .map(callExpr -> visitCalleeParent(callExpr, chain))
+        .orElse(chain);
+    }
+    return chain;
+  }
+
+
+  private List<QualifiedExpression> visitCalleeParent(CallExpression call, List<QualifiedExpression> chain) {
+    return Optional.of(call.callee())
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(QualifiedExpression.class))
+      .map(qe -> visitParent(qe, chain))
+      .orElse(chain);
+  }
+
+  private static Optional<CallExpression> ignoreSubscriptionAndGetCallExpr(SubscriptionExpression qualifier) {
+    if (qualifier.object().is(Tree.Kind.CALL_EXPR)) {
+      return Optional.of((CallExpression) qualifier.object());
+    } else if (qualifier.object().is(Tree.Kind.SUBSCRIPTION)) {
+      return ignoreSubscriptionAndGetCallExpr((SubscriptionExpression) qualifier.object());
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isValidPandasCall(List<QualifiedExpression> chain) {
+    QualifiedExpression firstQualifiedExpression = chain.get(chain.size() - 1);
```

Comment:
The assumption is that if the first call returns a data frame, every subsequent call will do that as well? Do we risk FPs with that assumption (Not challenging the assumption, just curious about potential edge case risk).

Reply:
Yes maybe you are right. It would be particular cases. For example:

```python
 pd.read_csv("some_csv.csv").axes[1].join(pd.Index([4, 5, 6])).repeat([1,2]).drop_duplicates().insert(1, 42)
```

Here we are working on an `Index` after the `axes[1]` instruction. We would raise an issue here. The problem is that there is no `pipe` method on `Index`, also changing the read_csv to pipe would be less clear. One way to circumvent this would be to check that at least the first two instructions return a `DataFrame`. It would make more sense than to refactor with a pipe method. That said I am not convinced this is a case we would encounter often. I am not 100% sure we would gain much from implementing this use case.  WDYT?

Reply:
I think we can accept this risk of FP. Maybe we can have such example documented in the tests, though?
So that we could revisit this later if this proves to be problematic.

Reply:
Okay so in the end we would have more FNs than FPs. As I did not implement the support for subscription with name we will not raise issue with such cases. But also we will not raise issues with the following case: 
```python
    pd.read_csv("some_csv.csv").filter(like='joe', axis=0).add(10).add(10).groupby("team")["salary"]["test"].axes[1].unique() # FN
```